### ISSUE TITLE
Loading.tsx and Suspense do not work on second page load and on

### ISSUE THREAD

**Issue by Vovch** [vercel/next.js#73474]
### Link to the code that reproduces this issue

https://github.com/Vovch/test-next-2

### To Reproduce

1. pnpm i
2. add .env file for DB connection (Postgres)
3. pnpm dev
4. navigate to localhost:3000/dashboard
5. reload the page
6. reload the page again
7. comment out or delete the Suspense wrapper in app\dashboard\page.tsx
8. reload the page
9. reload the page again

### Current vs. Expected behavior

Expected:
On steps 4 and 5 I expect the Suspense component to be rendered correctly without delaying the page load.
Steps 8 and 9 - Expected component from loading.tsx to be shown on the screen

Current: 
The page is not loaded into the browser at all before all the components are rendered on the server (3+ seconds). Suspense or loading.tsx components are not shown.

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Enterprise
  Available memory (MB): 32257  
  Available CPU cores: 8        
Binaries:
  Node: 18.20.3
  npm: 10.8.1
  Yarn: N/A
  pnpm: 9.11.0
Relevant Packages:
  next: 15.0.4-canary.36 // Latest available version is detected (15.0.4-canary.36).
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.5.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Lazy Loading

### Which stage(s) are affected? (Select all that apply)

next dev (local)
next start (local)

### Additional context

I was also able to reproduce this issue with next 15.0.3 and 14.2.18.

The code was created by following the tutorial up to and including chapter 9 https://nextjs.org/learn/dashboard-app/streaming and then trying to find the issue (since I couldn't make it work while following the tutorial).

in app\dashboard\page.tsx there is a commented out string with `await new Promise(...)`. If enabled it causes the app to work correctly. The effect depends on milliseconds value passed to the timeout, higher values (1000 and on) make the best result, lower values like `500` makes the defect reproducible again.

Intermittently I'm unable to get the loaders even when navigating between the pages by links in the app even if the first loaded page is not the dashboard itself.

**Comment by ayinde-xyz** [comment]
I had an error similar to this when trying to fetch dynamic data and I tried implementing the react Suspense component

**Comment by Vovch** [comment]
> @Vovch I tried deprecating my next.js version to 14.2.17 and the error disappeared.

Thanks for replying. Unfortunately, in my case the issue still persists even on 14.2.17
![Code_nvNKZwPoKO](https://github.com/user-attachments/assets/9a344436-6d48-4bf6-800a-5302afb5fdd8)

**Comment by ChongyuWang150** [comment]
Hi there,

I’m a student at the University of Michigan, and I’m currently working on a project for my coursework that involves contributing to an open-source repository. I have experience working with Next.js and would love to help address this issue.

The problem seems clear, and I’ve reviewed the reproduction case in the StackBlitz demo. Please let me know if I can go ahead and work on this or if there are any specific guidelines I should follow.

Looking forward to contributing!

Best,
Charlie

**Comment by Vovch** [comment]
@ChongyuWang150 https://github.com/vercel/next.js/blob/canary/contributing.md I believe here is the contribution guide.

**Comment by ElectricCodeGuy** [comment]
Loading.tsx does only work on the very first render of the route. If you want to re-trigger suspense you need to set the "key" prop in suspense and use what ever that are changing, such as the searchParams or Params as the key.

**Comment by Vovch** [comment]
> Loading.tsx does only work on the very first render of the route. If you want to re-trigger suspense you need to set the "key" prop in suspense and use what ever that are changing, such as the searchParams or Params as the key.

@ElectricCodeGuy thank you for your reply! Unfortunately this seems to be not the case, since Suspense loader is not shown even on initial page load, but only once for a single user after starting the server, and not even every time the server is started. 

Also the difference in its work with and without the Promise resolved on timeout shows that it might be a little more complex than just key assigned to the suspense element.

Just to make sure, I've checked the code with `Math.random()` as a key for Suspense - still no luck.

**Comment by Vovch** [comment]
@ElectricCodeGuy I'm sorry if my description of the issue was not clear. Please have a look at https://nextjs.org/learn/dashboard-app/streaming , especially the part with "Streaming a whole page with loading.tsx", I think the explanation there is better than mine. 

In my code you could also try uncommenting the promise  `// await new Promise((resolve) => setTimeout(resolve, 1000));` and see that suspense is working after that. You also might want to comment out the suspense wrapper on `RevenueChart` component and see that `loading.tsx` is shown on every page reload when the promise in the `app/dashboard/page.tsx` is not commented out and that it does is not shown when it is commented.

I came to this part by simply following the tutorial on the main page of Next.js: https://nextjs.org/learn

Here is the part about the loader:
![image](https://github.com/user-attachments/assets/611642d5-ca87-48ba-85cc-e63def4c71ac)

And here is the part about Suspense component:
![image](https://github.com/user-attachments/assets/efe595b4-d668-4f16-a4a5-6d8c3673aec1)

**Comment by ElectricCodeGuy** [comment]
> @ElectricCodeGuy I'm sorry if my description of the issue was not clear. Please have a look at https://nextjs.org/learn/dashboard-app/streaming , especially the part with "Streaming a whole page with loading.tsx", I think the explanation there is better than mine.
> 
> In my code you could also try uncommenting the promise `// await new Promise((resolve) => setTimeout(resolve, 1000));` and see that suspense is working after that. You also might want to comment out the suspense wrapper on `RevenueChart` component and see that `loading.tsx` is shown on every page reload when the promise in the `app/dashboard/page.tsx` is not commented out and that it is not shown when the promise is commented out.
> 
> I came to this part by simply following the tutorial on the main page of Next.js: https://nextjs.org/learn
> 
> Here is the part about the loader: ![image](https://private-user-images.githubusercontent.com/14871626/394282596-611642d5-ca87-48ba-85cc-e63def4c71ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4NjgzNzQsIm5iZiI6MTczMzg2ODA3NCwicGF0aCI6Ii8xNDg3MTYyNi8zOTQyODI1OTYtNjExNjQyZDUtY2E4Ny00OGJhLTg1Y2MtZTYzZGVmNGM3MWFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDIyMDExNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2NzAzNjk1ZmZlZTU2YjE1MmIzOWJhNDUyZmY3YTJkYmQ4OGM1ZWJjZDIzNTk2YjkxNzk2YzQxNzE1OGQ4YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Sov55cxs1b_O9rKKs47NRyquH4frLTKyZBiFwmLmFrE)
> 
> And here is the part about Suspense component: ![image](https://private-user-images.githubusercontent.com/14871626/394283096-efe595b4-d668-4f16-a4a5-6d8c3673aec1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4NjgzNzQsIm5iZiI6MTczMzg2ODA3NCwicGF0aCI6Ii8xNDg3MTYyNi8zOTQyODMwOTYtZWZlNTk1YjQtZDY2OC00ZjE2LWE0YTUtNmQ4YzM2NzNhZWMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDIyMDExNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2M2RkYzE1ODA5MmZmYmY1ODE2NzRjMjZjMmQ3ZDcwN2ZjMThmNzgwMDBmODZiYTFlNGYwNGVkZGNiMzY3ZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.c350Tkdct3PVUlpooDqsx0eGELIzu1jXiNrlXRrGm7U)



```

import { generateYAxis } from '@/app/lib/utils';
import { CalendarIcon } from '@heroicons/react/24/outline';
import { lusitana } from '@/app/ui/fonts';
import { fetchRevenue } from '@/app/lib/data';
 
// ...
 
export default async function RevenueChart() { // Make component async, remove the props
  const revenue = await fetchRevenue(); // Fetch data inside the component
 
  const chartHeight = 350;
  const { yAxisLabels, topLabel } = generateYAxis(revenue);
 
  if (!revenue || revenue.length === 0) {
    return <p className="mt-4 text-gray-400">No data available.</p>;
  }
 
  return (
    // ...
  );
}

```

That's because a promise exist inside of it, so this promise would trigger the suspense :)

**Comment by Vovch** [comment]
@ElectricCodeGuy true, it should. My problem is the suspense is not triggered on second load after the server start and on. :) Or, even after navigating between the pages.

If you had a look at the code I provided, you would see that it is exactly what's going on:
https://github.com/Vovch/test-next-2/blob/c1015e473a9c738c78bbb3beb8749de5f593b2b5/app/ui/dashboard/revenue-chart.tsx#L16

https://github.com/Vovch/test-next-2/blob/c1015e473a9c738c78bbb3beb8749de5f593b2b5/app/lib/data.ts#L18

So, you're saying you were able to make it work like https://nextjs.org/learn/dashboard-app/streaming says, on the page reload or it is just a theory?

Anyway, thank you for trying.

**Comment by Bo-Teng** [comment]
You may take a look at this issue, it may solve your problem https://github.com/vercel/next.js/issues/69625

**Comment by Vovch** [comment]
> Have you found any solution to this problem?


Nope, switched to page router as it seems that it has much less issues. Will try to implement such things with Lazy Loading, when needed.

**Comment by joelvdavies** [comment]
I too have been having this issue after trying to use suspense. I pulled the next-learn repo and found exactly the same thing when I tried to run it. https://github.com/vercel/next-learn/issues/952. I have tried a few different machines node 20, node 22, and various options I found in the following stack overflow posts

https://stackoverflow.com/questions/77949807/suspense-blocks-the-whole-page-when-used-in-server-component-for-streaming
https://stackoverflow.com/questions/75947098/suspense-fallback-is-not-showing-in-nextjs-13/78473852#78473852
https://stackoverflow.com/questions/78929824/next-js-suspense-server-side-doesnt-work

but still haven't found a solution unfortunately. It appears it may have been cropping up for a while but is not very reproducible. (Although I cannot get it work myself)

**Comment by vadolasi** [comment]
Same error here

**Comment by rcocco** [comment]
It could be caused by security software, such as Kaspersky.
When I disabled Kaspersky's Web Anti-Virus, everything worked fine.
I guess Kaspersky's Web Anti-Virus caches responses with content-type: text/html or content-type: text/x-component for virus scanning.

**Comment by pedronastasi** [comment]
> It could be caused by security software, such as Kaspersky. When I disabled Kaspersky's Web Anti-Virus, everything worked fine. I guess Kaspersky's Web Anti-Virus caches responses with content-type: text/html or content-type: text/x-component for virus scanning.

It worked for me.. It is weird that Kaspersky does this.. 🤕

**Comment by CodingMeSwiftly** [comment]
I'm facing this issue in every browser on my MacBook Pro, 2023.

- I see the Suspense fallback installed by `loading.tsx` in dev mode (`next dev`). 
- In production (`next build && next start`), streaming seems to be broken and I see long "Waiting for server response" times.
- I don't have any security software installed. 
- All the pages in my test setup are statically rendered and I invoke a server action to clear the cache (`revalidatePath`) before navigating to a page that should show a suspense fallback, but doesn't. 
- The React component tree looks fine to me, Suspense switches `key` between navigation, so there's no issue there.
- <img width="190" height="64" alt="Image" src="https://github.com/user-attachments/assets/2261f40f-b64a-49e0-8f50-d300598e9b09" />
- `next.config.ts`:
```
import type { NextConfig } from 'next'

const nextConfig: NextConfig = {
    async headers() {
        return [
            {
                source: '/:path*{/}?',
                headers: [
                    {
                        key: 'X-Accel-Buffering',
                        value: 'no'
                    }
                ]
            }
        ]
    }
}

export default nextConfig
```

I'll try hosting my test setup on Vercel next, but this should really work in local prod mode, no?

**Comment by ardonjr** [comment]
Okay, this is super dumb, but here is an issue I ran into. Maybe it can help someone else who is stuck, so even though I'm fully ashamed, I'm posting this anyway.

I also ran into this issue where the loading page didn't show up while following along the nextjs learning ( https://nextjs.org/learn/dashboard-app/streaming#streaming-a-whole-page-with-loadingtsx )
Opening localhost in a different browser didn't work either. I stopped and restarted my Nextjs project. But the Dashboard page just hangs for the data request to complete before showing anything.

In Dev Tools under network I could see the request for the Dashboard page just hangs for 3.28 seconds before doing anything else.

It took me a while but I finaly found out that I had placed the 'loading.tsx' page in '/app/ui/dashboard' instead of '/app/dashboard' 
Moving the page to the correct folder '/app/dashboard' obviously fixed the issue for me.

**Comment by hrishikesh2810** [comment]
Hi maintainers 👋
I’d like to work on this issue.
Please let me know if it’s already assigned to someone.
Thanks!

**Comment by gonzoblasco** [comment]
I have been going through the comments and the related issues referenced here, and I think there are a few distinct scenarios being conflated in this thread. Separating them might help narrow down the root cause.

**Scenario A — Antivirus/proxy interference (Kaspersky)**
Several users reported that disabling Kaspersky Web Anti-Virus resolved the issue. @rcocco identified that Kaspersky appears to cache responses with `content-type: text/html` or `content-type: text/x-component` for virus scanning, which would block the streaming chunks from arriving incrementally. This is an environment issue, not a Next.js bug — but it is worth documenting because it affects many users who follow the official Next.js tutorial.

**Scenario B — Streaming works in dev but not in production (`next start`)**
@CodingMeSwiftly reported that `loading.tsx` works in `next dev` but not after `next build && next start`. This is a different problem. In production, if the response is being buffered by a reverse proxy (nginx, Cloudflare, etc.) or if `X-Accel-Buffering` is not set, the streaming boundary may not reach the client until the entire RSC payload is ready. The `X-Accel-Buffering: no` header they added is the right approach for nginx, but it would not help with other proxies.

**Scenario C — The `await` before Suspense "fix"**
The OP found that adding `await new Promise((resolve) => setTimeout(resolve, 1000))` before the Suspense boundary makes it work. This is the most interesting clue. If the parent page has no async work of its own, the RSC payload for the page and its Suspense boundary may be emitted in the same flush — meaning the browser receives the entire page content before the fallback has a chance to render. An artificial delay forces a separate flush, which lets the fallback show. This suggests the issue might be in how RSC flushes boundaries when the parent component has no async operations of its own.

**Related issues that seem relevant:**
- #69625 (mentioned by @Bo-Teng) — appears to be about the same streaming behavior
- #70763 (discussion, mentioned by @joelvdavies) — covers RSC streaming edge cases

It would help to know: has anyone reproduced this with a minimal test case that does **not** involve Kaspersky or a reverse proxy? And has anyone tested with `experimental.ppr` enabled, which changes how boundaries are streamed?

**Comment by marcoshernanz** [comment]
Thanks for the report. I tested the reproduction's `loading.tsx` and nested Suspense flow on the reported 15.0.4 canary and current canary, in both development and production. Next.js sent the fallback immediately on every initial request and reload. I could reproduce the failure only when the response was buffered, which matches the Kaspersky and Sophos findings here and in #70763 and #80693. Since this is downstream response buffering rather than a Next.js failure, I'm closing it. For self-hosting, ensure the full request path passes streaming responses through without buffering: https://nextjs.org/docs/app/guides/self-hosting#streaming-and-suspense
